File: app/src/main/java/org/thoughtcrime/securesms/notifications/RemoteReplyReceiver.java
/*
 * Copyright (C) 2016 Open Whisper Systems
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.thoughtcrime.securesms.notifications;

import android.annotation.SuppressLint;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;

import androidx.core.app.RemoteInput;

import org.session.libsession.messaging.messages.ExpirationConfiguration;
import org.session.libsession.messaging.messages.signal.OutgoingMediaMessage;
import org.session.libsession.messaging.messages.signal.OutgoingTextMessage;
import org.session.libsession.messaging.messages.visible.VisibleMessage;
import org.session.libsession.messaging.sending_receiving.MessageSender;
import org.session.libsession.snode.SnodeAPI;
import org.session.libsession.utilities.Address;
import org.session.libsession.utilities.recipients.Recipient;
import org.session.libsignal.utilities.Log;
import org.thoughtcrime.securesms.ApplicationContext;
import org.thoughtcrime.securesms.database.MarkedMessageInfo;
import org.thoughtcrime.securesms.database.MmsDatabase;
import org.thoughtcrime.securesms.database.SmsDatabase;
import org.thoughtcrime.securesms.database.Storage;
import org.thoughtcrime.securesms.database.ThreadDatabase;
import org.thoughtcrime.securesms.mms.MmsException;

import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;
import network.loki.messenger.libsession_util.util.ExpiryMode;

/**
 * Get the response text from the Wearable Device and sends an message as a reply
 */
@AndroidEntryPoint
public class RemoteReplyReceiver extends BroadcastReceiver {

  public static final String TAG           = RemoteReplyReceiver.class.getSimpleName();
  public static final String REPLY_ACTION  = "network.loki.securesms.notifications.WEAR_REPLY";
  public static final String ADDRESS_EXTRA = "address";
  public static final String REPLY_METHOD  = "reply_method";

  @Inject
  ThreadDatabase threadDatabase;
  @Inject
  MmsDatabase mmsDatabase;
  @Inject
  SmsDatabase smsDatabase;
  @Inject
  Storage storage;

  @SuppressLint("StaticFieldLeak")
  @Override
  public void onReceive(final Context context, Intent intent) {
    if (!REPLY_ACTION.equals(intent.getAction())) return;

    Bundle remoteInput = RemoteInput.getResultsFromIntent(intent);

    if (remoteInput == null) return;

    final Address      address      = intent.getParcelableExtra(ADDRESS_EXTRA);
    final ReplyMethod  replyMethod  = (ReplyMethod) intent.getSerializableExtra(REPLY_METHOD);
    final CharSequence responseText = remoteInput.getCharSequence(DefaultMessageNotifier.EXTRA_REMOTE_REPLY);

    if (address     == null) throw new AssertionError("No address specified");
    if (replyMethod == null) throw new AssertionError("No reply method specified");

    if (responseText != null) {
      new AsyncTask<Void, Void, Void>() {
        @Override
        protected Void doInBackground(Void... params) {
          Recipient recipient = Recipient.from(context, address, false);
          long threadId = threadDatabase.getOrCreateThreadIdFor(recipient);
          VisibleMessage message = new VisibleMessage();
          message.setSentTimestamp(SnodeAPI.getNowWithOffset());
          message.setText(responseText.toString());
          ExpirationConfiguration config = storage.getExpirationConfiguration(threadId);
          ExpiryMode expiryMode = config == null ? null : config.getExpiryMode();

          long expiresInMillis = expiryMode == null ? 0 : expiryMode.getExpiryMillis();
          long expireStartedAt = expiryMode instanceof ExpiryMode.AfterSend ? message.getSentTimestamp() : 0L;
          switch (replyMethod) {
            case GroupMessage: {
              OutgoingMediaMessage reply = OutgoingMediaMessage.from(message, recipient, Collections.emptyList(), null, null, expiresInMillis, 0);
              try {
                mmsDatabase.insertMessageOutbox(reply, threadId, false, null, true);
                MessageSender.send(message, address);
              } catch (MmsException e) {
                Log.w(TAG, e);
              }
              break;
            }
            case SecureMessage: {
              OutgoingTextMessage reply = OutgoingTextMessage.from(message, recipient, expiresInMillis, expireStartedAt);
              smsDatabase.insertMessageOutbox(threadId, reply, false, System.currentTimeMillis(), null, true);
              MessageSender.send(message, address);
              break;
            }
            default:
              throw new AssertionError("Unknown Reply method");
          }

          List<MarkedMessageInfo> messageIds = threadDatabase.setRead(threadId, true);

          ApplicationContext.getInstance(context).messageNotifier.updateNotification(context);
          MarkReadReceiver.process(context, messageIds);

          return null;
        }
      }.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
    }
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/notifications/RemoteReplyReceiver.kt
package org.thoughtcrime.securesms.notifications

import android.annotation.SuppressLint
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import androidx.core.app.RemoteInput
import androidx.core.content.IntentCompat
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import network.loki.messenger.R
import org.session.libsession.messaging.messages.applyExpiryMode
import org.session.libsession.messaging.messages.signal.OutgoingTextMessage
import org.session.libsession.messaging.messages.visible.VisibleMessage
import org.session.libsession.messaging.sending_receiving.MessageSender
import org.session.libsession.network.SnodeClock
import org.session.libsession.utilities.Address
import org.session.libsignal.utilities.Log
import org.thoughtcrime.securesms.database.MmsDatabase
import org.thoughtcrime.securesms.database.RecipientRepository
import org.thoughtcrime.securesms.database.SmsDatabase
import org.thoughtcrime.securesms.database.Storage
import org.thoughtcrime.securesms.database.ThreadDatabase
import org.thoughtcrime.securesms.database.getOrCreateThreadIdFor
import org.thoughtcrime.securesms.database.model.MessageId
import org.thoughtcrime.securesms.dependencies.ManagerScope
import org.thoughtcrime.securesms.pro.ProStatusManager
import javax.inject.Inject

/**
 * Get the response text from the notification and sends an message as a reply
 */
@AndroidEntryPoint
class RemoteReplyReceiver : BroadcastReceiver() {
    @Inject
    lateinit var threadDatabase: ThreadDatabase

    @Inject
    lateinit var smsDatabase: SmsDatabase

    @Inject
    lateinit var storage: Storage

    @Inject
    lateinit var clock: SnodeClock

    @Inject
    lateinit var recipientRepository: RecipientRepository

    @Inject
    lateinit var messageSender: MessageSender

    @Inject
    lateinit var proStatusManager: ProStatusManager

    @Inject
    @ManagerScope
    lateinit var scope: CoroutineScope

    @SuppressLint("StaticFieldLeak")
    override fun onReceive(context: Context, intent: Intent) {
        if (REPLY_ACTION != intent.action) return
        val remoteInput = RemoteInput.getResultsFromIntent(intent) ?: return

        val threadAddress = requireNotNull(
            IntentCompat.getParcelableExtra(
                intent,
                ADDRESS_EXTRA,
                Address.Conversable::class.java
            )
        ) {
            "Address must be specified"
        }

        val responseText = remoteInput.getCharSequence(REPLY_TEXT)

        if (responseText != null) {
            val r = goAsync()

            scope.launch {
                try {
                    val threadRecipient = recipientRepository.getRecipientSync(threadAddress)

                    val message = VisibleMessage()
                    message.sentTimestamp = clock.currentTimeMillis()
                    message.text = responseText.toString()
                    proStatusManager.addProFeatures(message)
                    message.applyExpiryMode(threadRecipient)

                    message.id = MessageId(
                        smsDatabase.insertMessageOutbox(
                            threadDatabase.getOrCreateThreadIdFor(threadAddress),
                            OutgoingTextMessage(
                                message,
                                threadAddress,
                                threadRecipient.expiryMode.expiryMillis,
                                0L
                            ),
                            false,
                            clock.currentTimeMillis()
                        ),
                        false
                    )

                    messageSender.send(message, threadAddress)

                    // By sending a message, we'll update the last seen also
                    storage.updateConversationLastSeenIfNeeded(
                        threadAddress = threadAddress,
                        lastSeenTime = clock.currentTimeMillis()
                    )
                } catch (e: Throwable) {
                    Log.e(TAG, "Error sending reply", e)
                } finally {
                    r.finish()
                }
            }
        }
    }

    companion object {
        private const val TAG: String = "RemoteReplyReceiver"
        private const val REPLY_ACTION: String = "network.loki.securesms.notifications.WEAR_REPLY"
        private const val REPLY_TEXT: String = "extra_remote_reply"
        private const val ADDRESS_EXTRA: String = "address"

        fun buildIntent(
            context: Context,
            threadAddress: Address.Conversable
        ): Pair<PendingIntent, RemoteInput> {
            // Reply Action
            val remoteInput = RemoteInput.Builder(REPLY_TEXT)
                .setLabel(context.getString(R.string.reply))
                .build()

            val replyIntent = Intent(context, RemoteReplyReceiver::class.java).apply {
                action = REPLY_ACTION
                putExtra(ADDRESS_EXTRA, threadAddress)
            }

            return PendingIntent.getBroadcast(
                context,
                1,
                replyIntent,
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_MUTABLE
            ) to remoteInput
        }
    }
}
